File: aribon/2-1/bubunwa.cpp
#include <iostream>
#include <cstdio>
#define MAX_N 20

// input
int a[MAX_N];
int n, k;

// iまででsumを作って、残りi以降を調べる
bool dfs(int i, int sum){
    // n個決め終わったら、今までの和sumがkと等しいかを返す
    if (i == n) return sum == k;

    // a[i]を使わない場合
    if (dfs(i+1, sum)) return true;

    // a[i]を使う場合
    if (dfs(i+1, sum+a[i])) return true;

    // a[i]を使う使わないにかかわらずkが作れないのでfalseを返す
    return false;
}

void solve(){
    if (dfs(0, 0)) printf("Yes\n");
    else printf("No\n");
}

File: aribon/2-1/bubunwa.py
# -*- coding:utf-8 -*-
import sys


def solve():
    """状態数が2^{n+1}程度なので、計算量はO(2^n)"""
    n = int(input())
    As = list(map(int, sys.stdin.readline().split()))
    k = int(input())

    def dfs(i, total):
        if total == k: return True

        if i == len(As): return False

        # As[i]を使う場合
        if dfs(i+1, total+As[i]): return True

        # As[i]を使わない場合
        if dfs(i+1, total): return True

        return False

    if dfs(0, 0):
        print("Yes")
    else:
        print("No")

if __name__ == "__main__":
    solve()
